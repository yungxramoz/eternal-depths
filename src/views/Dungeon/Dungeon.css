.encounter-img {
  /* width: 50vw; */
  filter: drop-shadow(-2px 2px 5px black);
}

.encounter-container {
  display: flex;
  flex-grow: 1;
  padding: 3em 0 3em 0;
}

.damage-animation {
  animation: damage 0.5s ease infinite;
}

.attack-animation {
  animation: attack 0.5s;
}

.idle-animation {
  animation: idle 1s ease infinite;
}

@keyframes damage {
  0% {
    transform: rotate(3deg);
  }
  25% {
    filter: grayscale(1) drop-shadow(-2px 2px 5px black);
  }
  50% {
    filter: grayscale(0.3) drop-shadow(-2px 2px 5px black);
  }
}

@keyframes attack {
  0% {
    transform: scale(1.1);
  }
}

@keyframes idle {
  0% {
    transform: translate(0px);
  }
  50% {
    transform: translateY(3px);
    filter: drop-shadow(-2px -1px 5px black);
  }
  100% {
    transform: translate(0px);
  }
}
