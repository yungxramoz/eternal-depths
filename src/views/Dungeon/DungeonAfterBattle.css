.action-buttons-container {
  display: flex;
  flex-grow: 1;
  flex-direction: column;
  justify-content: end;
}
