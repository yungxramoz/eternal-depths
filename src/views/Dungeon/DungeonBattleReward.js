import React from 'react'

const Reward = () => {
  handleOption1 = () => {
    // Handle the logic for option 1 here
  }

  handleOption2 = () => {
    // Handle the logic for option 2 here
  }
  return (
    <div>
      <h2>Choose your reward:</h2>
      <button onClick={this.handleOption1}>Option 1</button>
      <button onClick={this.handleOption2}>Option 2</button>
    </div>
  )
}

export default Reward
