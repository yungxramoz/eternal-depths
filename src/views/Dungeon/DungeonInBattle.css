.combat-control-container {
    display: flex;
    flex-direction: row;
    justify-content: space-between;
    align-items: flex-start;
    padding-top: 1rem;
    column-gap: 2rem;
}