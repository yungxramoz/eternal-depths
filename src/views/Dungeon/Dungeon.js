import React, { useState } from 'react'
import RpgButton from '../../components/atoms/RpgButton/RpgButton'
import RpgContainer from '../../components/templates/RpgContainer/RpgContainer'
import Encounter from '../../models/Encounter'
import './Dungeon.css'
import RpgProgressBar from '../../components/atoms/RpgProgressBar/RpgProgressBar'
import HpProgressBar from '../../components/molecules/ProgressBar/HpProgressBar'

const Dungeon = () => {
  let [encounter, setEncounter] = useState(new Encounter(1))

  const regenerateEncounter = () => {
    const randomLevel = Math.floor(Math.random() * 10) + 1
    const isBoss = Math.random() < 0.5
    setEncounter(new Encounter(randomLevel, isBoss))
  }
  let imgSrc = require('../../assets/images/encounters/' + encounter.fileName)

  let randomRoomType = Math.random() < 0.5 ? 'dungeon' : 'cave'
  let randomDungeonRoom = require(`../../assets/images/rooms/${randomRoomType}-${Math.floor(
    Math.random() * 5,
  )}.png`)

  return (
    <RpgContainer fullPage bgImg={randomDungeonRoom}>
      <h1>
        {encounter.name} Lvl {encounter.level}
      </h1>
      <h3>Stage: 1</h3>
      <HpProgressBar currentHp={encounter.hp-10} maxHp={encounter.maxHp} />
      <div className="encounter-container">
        <img
          className="encounter-img"
          src={imgSrc}
          alt={encounter.name}
          style={encounter.style}
        />
      </div>
      <RpgButton text="Regenerate" onClick={regenerateEncounter} />
    </RpgContainer>
  )
}

export default Dungeon
