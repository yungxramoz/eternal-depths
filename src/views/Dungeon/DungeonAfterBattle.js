import React from 'react'
import { useDispatch, useSelector } from 'react-redux'
import RpgButton from '../../components/atoms/RpgButton/RpgButton'
import { recoverHp } from '../../store/character/characterSlice'
import { gameWon, nextStage } from '../../store/game/gameSlice'
import './DungeonAfterBattle.css'

const DungeonAfterBattle = () => {
  const dispatch = useDispatch()
  const stage = useSelector((state) => state.game.stage)

  const localNextStage = () => {
    dispatch(nextStage())
    dispatch(recoverHp(15))
  }

  return (
    <>
      <h1>Stage {stage}</h1>
      <div className="action-buttons-container">
        <RpgButton onClick={() => dispatch(gameWon())} text="Escape" />
        <RpgButton onClick={localNextStage} text="Next Stage" />
      </div>
    </>
  )
}

export default DungeonAfterBattle
