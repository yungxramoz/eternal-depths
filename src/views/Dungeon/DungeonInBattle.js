import React, { useEffect } from 'react'
import { useDispatch, useSelector } from 'react-redux'
import AttackButton from '../../components/molecules/AtackButton/AttackButton'
import HpProgressBar from '../../components/molecules/ProgressBar/HpProgressBar'
import {
  attackEffects,
  calculatedCharacterStats,
  damageCharacter,
} from '../../store/character/characterSlice'
import {
  animateIdle,
  attack,
  battleDefeat,
  damageEncounter,
} from '../../store/game/gameSlice'
import { calculateDamage } from '../../utils/attack'
import './DungeonInBattle.css'
import { EMPTY_ATTACK } from '../../constants/attack-type'

const DungeonInBattle = ({ children }) => {
  const dispatch = useDispatch()
  const character = useSelector((state) => state.character.current)
  const encounter = useSelector((state) => state.game.encounter)
  const encounterTurn = useSelector((state) => state.game.encounterTurn)
  const characterStats = useSelector(calculatedCharacterStats)

  useEffect(() => {
    setTimeout(() => {
      if (encounterTurn) {
        if (encounter.hp > 0) {
          dispatch(attack())
          const damage = calculateDamage(
            EMPTY_ATTACK,
            encounter.stats,
            encounter.minDamage,
            encounter.maxDamage,
            characterStats,
          )
          dispatch(damageCharacter(damage))
        }
      } else {
        dispatch(animateIdle())
      }
    }, 500)
  }, [encounterTurn, dispatch, encounter, character.hp, characterStats])

  useEffect(() => {
    if (character.hp <= 0) {
      dispatch(battleDefeat())
    }
  }, [character.hp, dispatch])

  const invokeAttack = (attack) => {
    const { minDamage, maxDamage } = character.items.weapon.stats
    const damage = calculateDamage(
      attack,
      characterStats,
      minDamage,
      maxDamage,
      encounter.stats,
    )
    dispatch(damageEncounter(damage))
    dispatch(attackEffects({ attack, dealtDamage: damage }))
  }

  return (
    <>
      {children}
      <HpProgressBar
        prefix="Your "
        currentHp={character.hp}
        maxHp={character.maxHp}
      />
      <div className="combat-control-container">
        {character.attacks.map((attack) => (
          <AttackButton
            key={attack.id}
            disabled={attack.currentCooldown > 0 || encounterTurn}
            attack={attack}
            text={attack.name}
            onClick={() => invokeAttack(attack)}
          ></AttackButton>
        ))}
      </div>
    </>
  )
}

export default DungeonInBattle
