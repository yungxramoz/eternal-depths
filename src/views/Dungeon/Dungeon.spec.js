import { configureStore } from '@reduxjs/toolkit'
import { render } from '@testing-library/react'
import React from 'react'
import { Provider } from 'react-redux'
import gameReducer, {
  battleReward,
  battleStart,
  battleVictory,
  characterLevelUp,
  gameStart,
} from '../../store/game/gameSlice'
import Dungeon from './Dungeon'

const store = configureStore({
  reducer: {
    game: gameReducer,
  },
})

jest.mock('./DungeonAfterBattle/DungeonAfterBattle', () => () => (
  <div>DungeonAfterBattle</div>
))
jest.mock('./DungeonBeforeBattle/DungeonBeforeBattle', () => () => (
  <div>DungeonBeforeBattle</div>
))
jest.mock('./DungeonInBattle/DungeonInBattle', () => () => (
  <div>DungeonInBattle</div>
))
jest.mock('./DungeonLevelUp/DungeonLevelUp', () => () => (
  <div>DungeonLevelUp</div>
))
jest.mock('./DungeonReward/DungeonReward', () => () => <div>DungeonReward</div>)
jest.mock(
  '../../components/organisms/CharacterHeader/CharacterHeader',
  () => () => <div>CharacterHeader</div>,
)

store.dispatch(gameStart())

describe('Dungeon', () => {
  it('renders without crashing', () => {
    render(
      <Provider store={store}>
        <Dungeon />
      </Provider>,
    )
    expect(document.body).toMatchSnapshot('Dungeon')
  })
  it('renders DungeonBeforeBattle component on encounter state', () => {
    render(
      <Provider store={store}>
        <Dungeon />
      </Provider>,
    )
    expect(document.body).toMatchSnapshot('DungeonBeforeBattle')
  })
  it('renders DungeonInBattle component on battle state', () => {
    store.dispatch(battleStart())
    render(
      <Provider store={store}>
        <Dungeon />
      </Provider>,
    )
    expect(document.body).toMatchSnapshot('DungeonInBattle')
  })
  it('renders DungeonAfterBattle component on battle victory state', () => {
    store.dispatch(battleVictory())
    render(
      <Provider store={store}>
        <Dungeon />
      </Provider>,
    )
    expect(document.body).toMatchSnapshot('DungeonAfterBattle')
  })

  it('renders DungeonLevelUp component on level up state', () => {
    store.dispatch(characterLevelUp())
    render(
      <Provider store={store}>
        <Dungeon />
      </Provider>,
    )
    expect(document.body).toMatchSnapshot('DungeonLevelUp')
  })
  it('renders DungeonReward component on reward state', () => {
    store.dispatch(battleReward())
    render(
      <Provider store={store}>
        <Dungeon />
      </Provider>,
    )
    expect(document.body).toMatchSnapshot('DungeonReward')
  })
})
