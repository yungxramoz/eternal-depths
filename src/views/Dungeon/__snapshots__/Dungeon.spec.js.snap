// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dungeon renders DungeonAfterBattle component on battle victory state: DungeonAfterBattle 1`] = `
<body>
  <div>
    <div
      class="rpgui-container container framed full-page scrollable dungeon-container"
      style="background-image: url(test-file-stub); background-repeat: no-repeat; background-size: cover; background-position: bottom;"
    >
      <h2>
        Stage 
        1
      </h2>
      <div>
        CharacterHeader
      </div>
      <div>
        DungeonAfterBattle
      </div>
    </div>
  </div>
</body>
`;

exports[`Dungeon renders DungeonBeforeBattle component on encounter state: DungeonBeforeBattle 1`] = `
<body>
  <div>
    <div
      class="rpgui-container container framed full-page scrollable dungeon-container"
      style="background-image: url(test-file-stub); background-repeat: no-repeat; background-size: cover; background-position: bottom;"
    >
      <h2>
        Stage 
        1
      </h2>
      <div>
        CharacterHeader
      </div>
      <div>
        DungeonBeforeBattle
      </div>
    </div>
  </div>
</body>
`;

exports[`Dungeon renders DungeonInBattle component on battle state: DungeonInBattle 1`] = `
<body>
  <div>
    <div
      class="rpgui-container container framed full-page scrollable dungeon-container"
      style="background-image: url(test-file-stub); background-repeat: no-repeat; background-size: cover; background-position: bottom;"
    >
      <div>
        DungeonInBattle
      </div>
    </div>
  </div>
</body>
`;

exports[`Dungeon renders DungeonLevelUp component on level up state: DungeonLevelUp 1`] = `
<body>
  <div>
    <div
      class="rpgui-container container framed full-page scrollable dungeon-container"
    >
      <div>
        CharacterHeader
      </div>
      <div>
        DungeonLevelUp
      </div>
    </div>
  </div>
</body>
`;

exports[`Dungeon renders DungeonReward component on reward state: DungeonReward 1`] = `
<body>
  <div>
    <div
      class="rpgui-container container framed full-page scrollable dungeon-container"
      style="background-image: url(test-file-stub); background-repeat: no-repeat; background-size: cover; background-position: bottom;"
    >
      <h2>
        Stage 
        1
      </h2>
      <div>
        CharacterHeader
      </div>
      <div>
        DungeonAfterBattle
      </div>
    </div>
  </div>
</body>
`;

exports[`Dungeon renders without crashing: Dungeon 1`] = `
<body>
  <div>
    <div
      class="rpgui-container container framed full-page scrollable dungeon-container"
      style="background-image: url(test-file-stub); background-repeat: no-repeat; background-size: cover; background-position: bottom;"
    >
      <h2>
        Stage 
        1
      </h2>
      <div>
        CharacterHeader
      </div>
      <div>
        DungeonBeforeBattle
      </div>
    </div>
  </div>
</body>
`;
